### PR TITLE
Add forward-search in shell (#571)

### LINKED ISSUE #571: Ctrl+S - Forward-search
**Is your feature request related to a problem? Please describe.**
When using `Ctrl+R` to use the backward-search you may accidentally skip the history result you wanted and you are forced to cancel the search and start from the beginning.

**Describe the solution you'd like**
Pressing `Ctrl+S` performs forward-search (like emacs/readline), allowing selecting the history result you just skipped.

**Describe alternatives you've considered**
💔 

**Additional context**
As an emacs user, this issue annoys me every time 🙂

By the way, there's something mapped in `Ctrl+S` that seems to cancel the search and move the cursor one character back when pressed out of the search-mode, what is it supposed to do?

### PR DESCRIPTION
**Pull request by borzacchiello**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the documentation and the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

The PR adds a forward-search to the rizin shell, allowing selecting the history skipped using the backward search (Ctrl-R).

I added it to `Ctrl-G` since `Ctrl-S` was already mapped to [_backspace_](https://github.com/rizinorg/rizin/blob/41b4da831eef5bcaeb980415b64f26df366d8045/librz/cons/dietline.c#L1508) command. I choose `Ctrl-G` only because isn't mapped and `G` is close to `R` on the keyboard, but maybe `Ctrl-S` can be remapped instead. Let me know what you prefer, and I will modify the commit accordingly!

**Test plan**

If I run the commands:
```
[0x000bfe24]> aa
[0x000bfe24]> ab
[0x000bfe24]> ac
[0x000bfe24]> ad
```
And I hit `Ctrl-R`, I write `a`, and I hit `Ctrl-R` two times, I get:
```
(reverse-i-search (a)): ab
```
Using the _forward-search_ I can go back to the skipped commands. E.g., hitting `Ctrl-G` one time I get:
```
(forward-i-search (a)): ac
```

**Closing issues**

closes #571

### CODE REVIEW

**Review by XVilka — APPROVED** [review]

**Review by ret2libc — CHANGES_REQUESTED** [review]
Hi and thanks for your contribution!! This is indeed a nice feature to have.

Since you mentioned ^S is already taken, I was looking at its code and found out:
```C
		case 19: // ^S -- backspace
			if (gcomp) {
				gcomp--;
			} else {
				__move_cursor_left();
			}
			break;
```

As you can see the cursor is moved to the left only if `!gcomp`. `gcomp` is the variable that says if we are searching the history or not. If you look at ^R you see that when gcomp is active, `gcomp_idx` is incremented to move the search backward. So `gcomp_idx` should be decremented to move the search forward.

I think `^S` was already written with the idea of having forward-search, however the code incorrectly decrements `gcomp` instead of `gcomp_idx`. I think fixing that should be enough to have properly working forward-search.

**Comment by borzacchiello** [comment]
> Hi and thanks for your contribution!! This is indeed a nice feature to have.
> 
> Since you mentioned ^S is already taken, I was looking at its code and found out:
> 
> ```c
> 		case 19: // ^S -- backspace
> 			if (gcomp) {
> 				gcomp--;
> 			} else {
> 				__move_cursor_left();
> 			}
> 			break;
> ```
> 
> As you can see the cursor is moved to the left only if `!gcomp`. `gcomp` is the variable that says if we are searching the history or not. If you look at ^R you see that when gcomp is active, `gcomp_idx` is incremented to move the search backward. So `gcomp_idx` should be decremented to move the search forward.
> 
> I think `^S` was already written with the idea of having forward-search, however the code incorrectly decrements `gcomp` instead of `gcomp_idx`. I think fixing that should be enough to have properly working forward-search.

Done! I modified the commit fixing `^S` and deleting `^G`.

**Comment by XVilka** [comment]
@ret2libc why 0.4.0? It doesn't break backward compatibility and basically a bugfix.

**Review by XVilka — APPROVED** [review]

File: librz/cons/dietline.c [hunk 1]
```diff
@@ -1995,11 +1998,19 @@ RZ_API const char *rz_line_readline_cb(RzLineReadCallback cb, void *user) {
...
+							}
 						}
 					}
 				}
-				printf("\r (reverse-i-search (%s)): %s\r", I.buffer.data, gcomp_line);
+				if (gcomp_is_rev) {
+					printf("\r (reverse-i-search (%s)): %s\r", I.buffer.data, gcomp_line);
+				} else {
+					printf("\r (forward-i-search (%s)): %s\r", I.buffer.data, gcomp_line);
+				}
```
Comment (ret2libc) [inline librz/cons/dietline.c]:
```suggestion
				const char *prompt = gcomp_is_rev ? "reverse-i-search" : "forward-i-search";
				printf("\r (%s (%s)): %s\r", prompt, I.buffer.data, gcomp_line);
```

**Review by ret2libc — CHANGES_REQUESTED** [review]
If I press many times ^S i think the index keeps increasing, thus I have to press many times ^R to go back to the previous search element. I think the index should stop increasing when you reached the last element you can search.

File: librz/cons/dietline.c (hunk 1 above)
Comment (borzacchiello) [reply librz/cons/dietline.c]: Done

**Comment by borzacchiello** [comment]
> If I press many times ^S i think the index keeps increasing, thus I have to press many times ^R to go back to the previous search element. I think the index should stop increasing when you reached the last element you can search.

Now it should be fine. The problem was that `gcomp_idx` became negative

**Review by ret2libc — APPROVED** [review]
Great!! Thanks a lot for this ;)

**Pull request merged**